### ISSUE TITLE
chap03_Linear_Regression_gradientTape_model2

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#28]
"""
(iris 이용하기)
tf.GradientTape + regression model
 -> 미분계수 자동 계산 -> model 최적화(최적의 기울기와 절편 update)
 - x변수 : 2~4칼럼
 - y변수 : 1컬럼
 - model 최적화 알고리즘 : Adam
"""
from sklearn.datasets import load_iris # dataset
from sklearn.model_selection import train_test_split # dataset split
from sklearn.metrics import mean_squared_error, r2_score # model 평가
import tensorflow as tf
tf.executing_eagerly() # 즉시 실행 test : ver2.x 사용중 

### 1. input/output 변수 정의 
iris = load_iris()
inputs = iris.data[:, 1:]
outputs = iris.data[:, 0]
inputs.shape # (150, 3)
outputs.shape # (150,)

tf.random.set_seed(123) # W,B seed값
x_train, x_test, y_train, y_test = train_test_split(inputs, outputs, test_size = 0.3, random_state = 123)
x_train.dtype # dtype('float64')

### 2. model : model 클래스
class Model(tf.keras.Model) : # tf.karas.Model 이용해서 자식클래스(부모클래스) 정의
    def __init__(self) : # 생성자 
        super().__init__() # 부모생성자 호출
        self.W = tf.Variable(tf.random.normal([3,1])) # 기울기(가중치)
        self.B = tf.Variable(tf.random.normal([1])) # 절편
    def call(self, inputs) : # 메서드 재정의
        return tf.matmul(tf.cast(inputs, tf.float32),self.W) + self.B # 회귀방정식(예측치)
    -하나는 32 비트고, 하나는 64비트이면 오류발생하니까

-요런식으로 5번에서 모델 만들 것이야.
-model = Model()
-model.call(inputs) : call은 생략가능
-model(inputs)

### 3. 손실 함수
def loss(model, inputs, outputs):
  err = model(inputs) - outputs
  return tf.reduce_mean(tf.square(err)) # MSE

### 4. 미분계수(기울기) 계산  
def gradient(model, inputs, outputs) : 
    with tf.GradientTape() as tape:
        loss_value = loss(model, inputs, outputs) # 손실함수 호출  
        grad = tape.gradient(loss_value, [model.W, model.B])
        # 미분계수 -> 기울기와 절편 업데이트
    return grad # 업데이트 결과 반환

### 5. model 생성
model = Model() # 생성자

### 6. model 최적화 객체
opt = tf.keras.optimizers.Adam(learning_rate = 0.01)
-학습 전 초기손실값과 w,b
print("초기 손실값 : {:.6f}".format(loss(model, x_train, y_train)))
print("w : {}, b : {}".format(model.W.numpy(), model.B.numpy()))

### 7. 반복학습 : train
for step in range(300) :
    grad = gradient(model, x_train, y_train) # 기울기 계산
    -기울기 -> 최적화 객체 반영
    opt.apply_gradients(zip(grad, [model.W, model.B]))
    
    if (step+1) % 20 ==0 :
        print("step = {}, loss = {:.6f}".format((step+1), loss(model, inputs, outputs)))

- model 최적화
print("최종 손실값 : {:.6f}".format(loss(model, x_train, y_train)))
print("w : {}, b : {}".format(model.W.numpy(), model.B.numpy()))

- model test
y_pred = model.call(x_test)
-print(y_pred.numpy())

mse = mean_squared_error(y_test, y_pred)
print("mse = ", mse)

r2 = r2_score(y_test, y_pred)
print("r2 =", r2)

'''
최종 손실값 : 1.705331
w : [[ 0.7612573 ]
 [-0.20238313]
 [ 1.1611581 ]], b : [1.9997358]
mse =  1.0807632023184397
r2 = -0.349987345292778
'''
